### PR TITLE
Programs can create streams/datasets

### PR DESCRIPTION
**Pull request by gokulavasan**
Changes in Programmatic API to allow creation of stream and datasets in Program's configure method.

Note: 
1. Validation of whether dataset/stream is added with the same name but different properties is not part of this PR. Opening this PR to get feedback on the approach and public API changes.
2. Workflow (and WorkflowAction) can't create streams/datasets. It is blocked because of this: https://issues.cask.co/browse/CDAP-2956

JIRA : https://issues.cask.co/browse/CDAP-2777

Build : http://builds.cask.co/browse/CDAP-DUT2167-4

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/ProgramConfigurer.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+package co.cask.cdap.api;
+
+import java.util.Map;
+
+/**
+ * Basic Program Configurer.
+ */
+public interface ProgramConfigurer {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/ProgramConfigurer.java]: Is it specific to Program? Flowlet configurer also fits for extending from this interface, yet it's not a program (Flow is a program)

Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/ProgramConfigurer.java]: But we also have FlowletProgramRunner :)

**Comment by gokulavasan** [comment]
In this PR, the programs delegate the addition of datasets, streams to ApplicationConfigurer and thus the ApplicationSpecification contains all the streams, datasets information and the ProgramSpecification doesn't. Other option is to add these streams, datasets info to individual ProgramSpecifications.

**Comment by gokulavasan** [comment]
Closing this PR. Will reopen for 3.2.

**Pull request closed without merging**